### ISSUE TITLE
Undefined reference to cv and boost in compile.sh

### ISSUE THREAD

**Issue by deworrall92** [classner/fertilized-forests#27]
Hi,
I've had to reinstall a few things on my computer (running Ubuntu 14.04) including fertilized-forests. I can configure everything all right and have all the prerequisites, but I keep getting the following errors during the compile.

```
scons: *** [build/release/Presentation/C++/custom] Error 1
scons: *** [build/release/Presentation/C++/overview] Error 1
scons: *** [build/release/Presentation/C++/regression] Error 1
libfertilized.so: undefined reference to `cv::String::allocate(unsigned long)'
libfertilized.so: undefined reference to `cv::imread(cv::String const&, int)'
libfertilized.so: undefined reference to `cv::String::deallocate()'
libfertilized.so: undefined reference to `boost::archive::detail::shared_ptr_helper::shared_ptr_helper()'
libfertilized.so: undefined reference to `boost::archive::detail::shared_ptr_helper::~shared_ptr_helper()'
libfertilized.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
scons: *** [build/release/fertilized_tests/fertilized_tests] Error 1
scons: building terminated because of errors.
```

I'm a bit confused since I already installed opencv and boost, and if I didn't properly beforehand then surely that would be corrected for me in setup_environment.sh? Do you know what I should to fix this?

**Comment by ChrislS** [comment]
Interesting! This should definitely not be  happening!
Could you post the contents of your generated compile.sh and setup_environment.sh? Bad timing, I am travelling for two weeks, but I'll try to help as good as possible. :)

**Comment by deworrall92** [comment]
Thanks for the help. Hope I'm not detracting from you travel time:

setup_environment.sh
```
#!/bin/bash
# Author: Christoph Lassner.

echo "This is the Linux environment setup script of the fertilized-forests library.

It is a guided walkthrough through the requirement installations and setup options. It will install required software and create/customize two files on success:

* setup_paths.sh
* compile.sh

You can use and modify them easily, since they are very short and simple to read. Use them to build the library afterwards.

This script only works on Debian systems (since it uses the 'apt-get' command)!
"

if [[ $1 == '--quiet' ]]; then
  echo "Quiet mode."
else
  echo "Press enter to continue."
  read a
fi

echo "Updating package cache..."
sudo apt-get update > /dev/null 2>&1
echo "Done."

##############################################
# Check for the availability of Python and pip

if [[ $9 != "" ]]; then
  echo "Using Python and Pip at:" $9
else
  if [[ `which python` != "" ]]; then
    echo "Using Python at: `which python`."
  else
    echo "No Python found on the command line! You can register an existing Python installation only for this script by running 'sudo PATH=\"/mypathtopython/bin:\$PATH\" ./setup_environment.sh' or by adding the line 'export PATH=\"/mypathtopython/bin:\$PATH\"' to your ~/.bashrc file permanently."

    if [[ $1 == '--quiet' ]]; then
      echo "Exiting..."
      exit 1
    else
      echo "
Alternatively, if you have no Python installed, should I install it by running 'apt-get install python python-dev python-pip' ('n' will exit)? [(y), n]"

  	read a
  	if [[ $a == "Y" || $a == "y" || $a = "" ]]; then
  	  echo "Installing..."
  	  sudo apt-get install python python-dev python-pip
  	  echo "Installation finished."
  	  if [[ `which python` == "" || `which pip` == "" ]]; then
  	    echo "Can not find python or pip on the command line. Installation apparently failed. You will have to fix this manually. Exiting..."
  	    exit 1
  	  fi
          else
            exit 1 
  	fi
    fi
  fi

  if [[ `which pip` != "" ]]; then
    echo "Using pip at: `which pip`."
  else
    if [[ $1 == '--quiet' ]]; then
      echo "Exiting..."
      exit 1
    else
      echo "No pip found! Should I install it by running 'apt-get install python-pip' ('n' will exit)? If you are using a Python distribution other than the system one, you will have to install it yourself! [(y), n]"

  	read a
  	if [[ $a == "Y" || $a == "y" || $a = "" ]]; then
  	  echo "Installing..."
  	  sudo apt-get install python-pip
  	  echo "Installation finished."
  	  if [[ `which python` == "" || `which pip` == "" ]]; then
  	    echo "Cannot find python or pip on the command line. Installation apparently failed. You will have to fix this manually. Exiting..."
  	    exit 1
  	  fi
          else
            exit 1 
  	fi
    fi
  fi
fi

################################
# Entering Python layer.
# Forwarding parameters.
# c.t. http://stackoverflow.com/questions/6962861/linux-shell-append-variable-parameters-to-command
for i in $@;
do
    params=" $params $i"
done
if [[ $9 != "" ]]; then
  $9/python setup_environment_linux.py $params
else
  python setup_environment_linux.py $params
fi
```

compile.sh
```
#!/bin/bash
echo You can speed up the build process by editing
echo "compile.sh" and increasing the --jobs=1
echo value.

echo Setting up paths...
source setup_paths.sh

echo Compiling...
# Add --with-python to build the python interface.
# Add --with-matlab to build the MATLAB interface.
# Add --with-tests to build the tests.
# Add --with-examples to build the examples.
# Add --with-caffe to build the CAFFE feature extraction.
# Configure --caffe-model-dir to change the location of the CAFFE models for feature extraction.
# Configure --temp-folder to change the CAFFE tmp folder.
# Configure --cuda-architectures=20;30;35;50 to change the available compiled graphics card architectures.
# Add --cpu-only to build the CPU only version of CAFFE.
# Add --with-checks to enable assertions.
# Add --disable-optimizations to create an -Od build for debugging.
scons --with-serialization --with-python --with-tests --with-examples --with-caffe --caffe-model-dir=../forestCaffe --jobs=12
```

**Comment by ChrislS** [comment]
Sorry, I meant setup_paths.sh. Could you also post that one? 
Does the core library build for you, i.e., if you `source setup_paths.sh` and then run `scons --with-serialization`, does the build succeed?

**Comment by deworrall92** [comment]
Yeah, the core library builds.

I tried adding some folder locations to `setup_paths.sh`, like the protobuf location and boost location, but they made no difference. Note sure if this is relevant, but I did have to alter some of the `.gitmodules` because mdb has been moved - I redirected to a mirror of the submodule.


 setup_paths.sh
```
#!/bin/bash
# Modify this file as required.
export BOOST_ROOT="/usr/include/boost"
# export BOOST_LIB_DIR=...
# export OPENCV_ROOT=...
export OPENCV_LIB_DIR=:"/usr/local/lib"
export PROTOBUF_ROOT="/usr/include/google/protobuf"
# export PROTOC=...
# export HDF5_ROOT=...
# export HDF5_LIB_DIR=...
# export HDF5_INCLUDE_DIR=...
# export PATH="...:$PATH"
# export OPENBLAS_ROOT=...

# export MATLAB_ROOT=...
export EIGEN_ROOT="/usr/include/eigen3"
```

**Comment by deworrall92** [comment]
woops, sorry, didn't mean to close the issue

**Comment by ChrislS** [comment]
Ok, looking all good (and btw., thanks for the info about the moved mdb!). On Linux, you should even be able to omit the OpenCV and Boost paths as long as you didn't install the libraries to custom locations. Hmm, strange. Unfortunately, I can currently not reproduce the error, since I don't have all necessary data with me and only occasionally very slow internet connection, etc. I think I'll have to make you wait until ~Aug. 25th, but I will look into this ASAP. Sorry for that! :-/

According to the error message, only the examples and the tests fail to build. Since all CI builds are fine from the latest stable (but I can't speak for the MDB updates) you should be fine without the tests for these two weeks.

If you link your own CPP projects against the library, you will have to link against OpenCV and Boost yourself anyway...

Best,
Christoph

**Comment by ChrislS** [comment]
Hi Daniel,

I'm back and investigating this issue. I immediately ran into the LMDB URL change and fixed this on the `unstable` branch. However, I can not reproduce the linking issue on my systems (CI build is still pending), running on Ubuntu 14.04. Are there any new findings from your side?

Best,
Christoph

**Comment by deworrall92** [comment]
Hey, unfortunately no new developments. I've actually side stepped using the Fertilized-Forest for the moment. Thanks anyway.

**Comment by classner** [comment]
Hi Daniel,

just an update on this: adressing many issues, among others the build system, I did a major update on version 1.0.2 and you'd be welcome to test it! The library is now using CMake and should be setup in Ubuntu in minutes (leaving aside some compile time ;) ). Sorry for the inconvenience with the SCons stuff before, I did my best to make it as convenient as possible, but I believe the CMake build system is a huge step forward.

Let me know what you think, otherwise, I am going to close this issue in a few days since it only concerns the former build system.
